### DISCUSSION TITLE
UMLS detection on detected entities

### DISCUSSION THREAD

**Discussion by kswamy15** [medspacy/medspacy#61] [Q&A]
Can we just use the spacy umls pipeline instead of trying to download and install UMLS files (which are like 40 GB in size)?  spacy says :
from scispacy.linking import EntityLinker
nlp.add_pipe("scispacy_linker", config={"resolve_abbreviations": True, "linker_name": "umls"})

Not sure what the limitations on this approach would be.  Just seeing if others have tried this.

**Comment by turbosheep** [comment]
Using the scispaCy UMLS linker does not avoid the download unfortunately. Their component will download the resources on the first run and store them in your user directory by default.

I have all the files downloaded and they are just over 2gb (this is ALL the dictionaries listed on their github page including rxnorm, hpo, etc. not just the UMLS files). If storage or downloads are a concern for you, scispaCy will definitely be easier than downloading metathesaurus files and generating quickUMLS resources, but it does not avoid the problem entirely.

**Reply by kswamy15** [reply]
Just want to make sure I understand the quickUMLS download for medspacy.  Is your 2GB file size refer to the sample concepts in medspacy?  Because I see like 35 GB size files in the UMLS website.  So want to make sure what steps I need to take to get UMLS concepts in medspacy beyond a few sample concepts.  

When I went to this page https://github.com/medspacy/medspacy/blob/master/windows_and_quickumls.md - it says 'By default, we package up a small set of SAMPLE concepts from UMLS and make them available as resources. To generate additional resources from scratch from UMLS, see the instructions at QuickUMLS.'  

Because of this, I went to QuickUMLS here https://github.com/Georgetown-IR-Lab/QuickUMLS.  There it says 'To install UMLS, you must first obtain a license from the National Library of Medicine; then you should download all UMLS files from this page; finally, you can install UMLS using the MetamorphoSys tool as explained in this guide. The installation can be removed once the system has been initialized.'  

Then I went to this page https://www.nlm.nih.gov/research/umls/licensedcontent/umlsknowledgesources.html.  Here I see three different files.  Full Release (umls-2020AB-full.zip), UMLS Metathesaurus Files, MRCONSO.RRF.  Only the .RRF file is 2 GB in size.  The other two files are like 35 GB in size.

**Reply by turbosheep** [reply]
Hi, sorry if I was unclear. It seems like you might be slightly confusing QuickUMLS and scispacy's EntityLinker. These are both spacy components and have almost identical functionality, but have different setup requirements.

The instructions you linked are correct for using QuickUMLS. To get more CUIs for quickUMLS, you will have to download the full release of UMLS which is around 35GB. We cannot distribute any larger pre-made resources due to UMLS licensing, even though we would eventually like to. 

Scispacy _does_ distribute a pre-made file with more UMLS concepts and  the EntityLinker will reach out to their servers to download those resources the first time you run your component. These files, created and distributed by the AllenAI team, are only about 2 gb.  It does not require downloading the full UMLS release.

The main difference between scispaCy's EntityLinker and QuickUMLS is the level of customization. You can do more to configure QuickUMLS to a specific project by changing matching conditions and resource files. scispaCy requires less manual setup but, as far as I know, does not explain how to create new resource files yourself in case what they distribute is inadequate.

Finally, one other thing I just noticed in your first code sample: 
```python
nlp.add_pipe("scispacy_linker", config={"resolve_abbreviations": True, "linker_name": "umls"})
```

This code is using the spacy 3.x syntax for adding components to a spacy pipeline. Scispacy is compatible with spaCy 3.x, but medspaCy and QuickUMLS are not yet compatible with 3.x. It is on our to-do list but we have not had the time internally to make the change. So if you chose to use QuickUMLS over scispaCy, please make sure to use spaCy 2.x.

I hope this helps clarify things.

**Reply by kswamy15** [reply]
This helps.  I didn't realize that scispacy s only compatible with spaCy 3.x.  So I may not be able to add the umls linker while using medspacy then.  I was thinking that I could use medspacy for sectionizer, ner, assertion and then use the umls from scispacy umls linker to identify the cui associated with the identified entities.  Based on your version comments, looks like this won't be possible.

**Reply by turbosheep** [reply]
Scispacy does have a 2.x compatible version. I believe scispacy 0.3.0 and earlier are all spacy 2.x. It is just no longer the default and we are running a bit behind with the updates 😄
